### PR TITLE
feat: add `extract_chain_id` method

### LINKED ISSUE #7893: Add bidirectional function around `v` signature calculations
### Describe the feature

As per https://github.com/paradigmxyz/reth/pull/7882#discussion_r1579753322, in the repo, we need to make calculations around `v` in `Signature` in both directions:

https://github.com/paradigmxyz/reth/blob/f6e1b0640cea24ed17efe2d4c09ceb591d6fa7cf/crates/primitives/src/transaction/signature.rs#L80-L95 

and in the other direction for the conversion included in https://github.com/paradigmxyz/reth/pull/7882

As a result, this would make sense to unify that into a standalone method to wrap this bidirectional conversion.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lakshya-sky**
Closes #7893

### CODE REVIEW

File: crates/primitives/src/transaction/signature.rs [hunk 1]
```diff
@@ -94,6 +94,22 @@ impl Signature {
         }
     }
 
+    /// Outputs (odd_y_parity, chain_id) from the `v` value.
+    /// This doesn't check validity of the `v` value for optimism.
+    #[inline]
+    pub fn inverse_v(v: u64) -> alloy_rlp::Result<(bool, Option<u64>)> {
```
Comment (Rjected) [inline crates/primitives/src/transaction/signature.rs:100]: let's rename this `extract_chain_id`

**Review by Rjected — CHANGES_REQUESTED** [review]
great start! i have a smol naming nit

File: crates/primitives/src/transaction/mod.rs [hunk 2]
```diff
@@ -1740,18 +1740,9 @@ impl TryFrom<reth_rpc_types::Transaction> for TransactionSignedEcRecovered {
...
-                                    // accordingly
-                                    chain_id.saturating_mul(2).saturating_add(35)
-                                } else {
-                                    // If CHAIN_ID is not available, set v = {0,1} + 27
-                                    27
-                                }))
-                            .is_zero()
+                            Signature::extract_chain_id(signature.v.to())
+                                .expect("v would always be valid")
+                                .0
```
Comment (Rjected) [inline crates/primitives/src/transaction/mod.rs:1745]: we can't panic here, this is a `TryFrom` impl, so we should return an error here

File: crates/primitives/src/transaction/signature.rs [hunk 3]
```diff
@@ -94,6 +94,22 @@ impl Signature {
...
+    pub fn extract_chain_id(v: u64) -> alloy_rlp::Result<(bool, Option<u64>)> {
+        if v < 35 {
+            // non-EIP-155 legacy scheme, v = 27 for even y-parity, v = 28 for odd y-parity
+            if v != 27 && v != 28 {
+                return Err(RlpError::Custom("invalid Ethereum signature (V is not 27 or 28)"))
+            }
+            Ok((v == 28, None))
+        } else {
+            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
+            Ok((((v - 35) % 2) != 0, Some((v - 35) >> 1)))
```
Comment (Rjected) [inline crates/primitives/src/transaction/signature.rs:109]: imo this was slightly more readable
```suggestion
            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
            let odd_y_parity = ((v - 35) % 2) != 0;
            let chain_id = (v - 35) >> 1;
            Ok((Self { r, s, odd_y_parity }, Some(chain_id)))
```

**Review by Rjected — CHANGES_REQUESTED** [review]
I have a few comments, otherwise this is almost there

File: crates/primitives/src/transaction/signature.rs (hunk 3 above)
Comment (lakshya-sky) [reply crates/primitives/src/transaction/signature.rs:109]: Why do we need `Self` on line 111?

Comment (Rjected) [reply crates/primitives/src/transaction/signature.rs]: We don't, just took it from the prev code, just noting that we want the variable names

File: crates/primitives/src/transaction/signature.rs [hunk 4]
```diff
@@ -94,6 +94,24 @@ impl Signature {
...
+                return Err(RlpError::Custom("invalid Ethereum signature (V is not 27 or 28)"))
+            }
+            Ok((v == 28, None))
+        } else {
+            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
+            let odd_y_parity = ((v - 35) % 2) != 0;
+            let chain_id = (v - 35) >> 1;
+            Ok((odd_y_parity, Some(chain_id)))
+        }
+    }
```
Comment (Rjected) [inline crates/primitives/src/transaction/signature.rs:113]: just noticed that this does not need to be a method on `Signature`

**Review by Rjected — CHANGES_REQUESTED** [review]
sorry, missed one thing - this can be a pure function that lives outside of the `Signature` struct

File: crates/primitives/src/transaction/signature.rs (hunk 4 above)
Comment (lakshya-sky) [reply crates/primitives/src/transaction/signature.rs:113]: I see, where do you suggest I should put the function?

Comment (Rjected) [reply crates/primitives/src/transaction/signature.rs:113]: In the same file would be fine

**Review by Rjected — APPROVED** [review]
looks good to me, thanks!

**Pull request merged**